src/s3.rs

```diff
@@ -11241,13 +11242,19 @@ impl<P, D> S3Client<P, D> where P: ProvideAwsCredentials, D: DispatchSignedReque
     /// Uploads a part in a multipart upload.
     /// **Note:** After you initiate multipart upload and upload one or more parts, you must either complete or abort multipart upload in order to stop getting charged for storage of the uploaded parts. Only after you either complete or abort multipart upload, Amazon S3 frees up the parts storage and stops charging you for the parts storage.
     pub fn upload_part(&self, input: &UploadPartRequest) -> Result<String, S3Error> {
-        let object_id = &input.key;
-        let mut request = SignedRequest::new("PUT", "s3", self.region, &format!("/{}", object_id));
+        let mut path = String::from("/");
+        if !is_dns_compatible(&input.bucket) {
+            path = path + &input.bucket + "/";
+        }
+        path = path + &input.key;
```

Comment:
There are other places in `+=` could be used in these changes, but I won't make more noise about them.